_pytest/main.py

```diff
@@ -468,6 +475,11 @@ def _parsearg(self, arg):
         if self.config.option.pyargs:
             parts[0] = self._tryconvertpyarg(parts[0])
         relpath = parts[0].replace("/", os.sep)
+        pattern_line = r':(?P<line>[0-9]+)$'
```

Comment:
Let's move all this logic into a separate free-function, this will separate that logic from the rest of this code and allow us to test the function independently:

```python
self._filter_by_function_name, relpath = _find_nearest_test_function(relpath)
```

```python
def _find_nearest_test_function(path_expr):
    """
    Finds the nearest test function given a path in the form:

    path/to/test.py:12

    With ``12`` being a line number, search upwards in the file 
    until it finds a test function declaration.

    :return: a tuple (function_name, real_path). function_name can be None in 
    case a function could not be found, and real_path will return the actual
    path stripping any trailing ``:<linnum>`` found.
    """    
```

Reply:
Fixed